packages/ui/src/elements/LeaveWithoutSaving/index.tsx

```diff
@@ -36,9 +60,47 @@ export const LeaveWithoutSaving: React.FC = () => {
     closeModal(modalSlug)
   }, [closeModal])
 
-  const onConfirm = useCallback(() => {
+  const onConfirm = useCallback(async () => {
+    const lockUser = documentLockState.current?.user
+
+    const isLockOwnedByCurrentUser =
+      typeof lockUser === 'object' ? lockUser?.id === user?.id : lockUser === user?.id
+
+    if (isLockingEnabled && documentIsLocked && (id || globalSlug)) {
+      // Check where user is trying to go
+      const nextHref = nextHrefRef.current
+      const nextPath = nextHref ? new URL(nextHref).pathname : ''
+
+      const isInternalView = ['/preview', '/api', '/versions'].some((path) =>
+        nextPath.includes(path),
+      )
+      // Only retain the lock if the user is still viewing the document
+      if (!isInternalView) {
+        if (isLockOwnedByCurrentUser) {
+          try {
+            await unlockDocument(id, collectionSlug ?? globalSlug)
+            setDocumentIsLocked(false)
+            setCurrentEditor(null)
+          } catch (err) {
+            console.error('Failed to unlock before leave', err)
+          }
+        }
+      }
+    }
+
     setHasAccepted(true)
-  }, [])
+  }, [
+    collectionSlug,
+    documentIsLocked,
+    documentLockState,
+    globalSlug,
+    id,
+    isLockingEnabled,
+    setCurrentEditor,
+    setDocumentIsLocked,
+    unlockDocument,
+    user?.id,
+  ])
```

Comment:
I think this component should take a callback instead of putting all of this logic in here. Leaving it as-is makes the component less re-usable.